pkg/process/net/network_linux.go

```diff
@@ -54,7 +56,7 @@ func GetRemoteSystemProbeUtil() (*RemoteSysProbeUtil, error) {
 		return nil, fmt.Errorf("remote tracer has no socket path defined")
 	}
 
-	if globalUtil == nil {
+	globalUtilOnce.Do(func() {
```

Comment:
Although this guarantees that only one thread will execute the `Once` closure, it does not guarantee anything about ordering.

If two threads arrive here at exactly the same instant, it is possible that one will execute the initialization code and the other will proceed past it, attempting to use `globalUtil` which may or may not have been initialized yet.

Reply:
If two threads are invoking `func (o *Once) Do(f func())` at the same time, only one of them will execute `f()` while the other will wait for the `f()` execution done by the other to complete.

That’s my interpretation of the following sentence extracted from the [`func (o *Once) Do(f func())` documentation](https://golang.org/pkg/sync/#Once.Do):

> Because **no call to Do returns until the one call to f returns**, if f causes Do to be called, it will deadlock.

This is confirmed by the [implementation source code of `func (o *Once) Do(f func())`](https://golang.org/src/sync/once.go?s=1473:1500#L30) which has a comment that describes precisely the situation that you are describing:

```
	// Note: Here is an incorrect implementation of Do:
	//
	//	if atomic.CompareAndSwapUint32(&o.done, 0, 1) {
	//		f()
	//	}
	//
	// Do guarantees that when it returns, f has finished.
	// This implementation would not implement that guarantee:
	// given two simultaneous calls, the winner of the cas would
	// call f, and the second would return immediately, without
	// waiting for the first's call to f to complete.
	// This is why the slow path falls back to a mutex, and why
	// the atomic.StoreUint32 must be delayed until after f returns.
```

So, `Do()`, guarantees that, when it returns, the closure has been fully executed, be it by the current thread or another one. So that it is guaranteed that `globalUtil` has been assigned by the closure before being dereferenced after the `Do()` call even in other threads.

Reply:
You're right, I missed that bit in the docs about callers blocking on `Do` until the initial closure is completed